RPC test port binding failure mitigation

awemany opened the pull request:
See commit messages. Meant to make the RPC tests robust against port binding failure with the resulting test failures and indefinite hangs.

awemany committed:
Add option that fails if not all P2P and RPC ports can be bound

Currently, bitcoind will attempt to bind both an IPv4 as well as an IPv6
TCP socket by default. If any of those fail, bitcoind will still come up
unless all bind()s fail.

This creates a couple of issues with automatic RPC testing. For example,
on a system that is *not* Travis (and thus reduces the ports it will bind to
a set of one each, as it won't bind to IPv6 localhost and thus will fail
properly for any binding failure), anything
that will block a port used by RPC testing, even if just temporarily so,
can cause a hang in the RPC test.

This can also be caused by parallelized running of tests with portseeds
 that cause ports to get in the way of each other, or when running multiple
RPC tests on a single machine.

This commit adds a '-bindallorfail' option, which will cause bitcoind
to fail during startup if any of the P2P or RPC TCP ports cannot be bound,
thus indicating cleanly to the test runner (manually or automatic through
rpc-tests.py) what went wrong.

awemany committed:
RPC testing: Auto assignment of new ports in case of failure

There were no provisions in the RPC code in case port binding of
either the RPC or P2P failed during any of the RPC tests. This has
caused hangs (without the fix in the previous commit to fail when
binding both the IPv4 and IPV6 localhost RPC or P2P port) in RPC
testing and is likely also reason for other sporadic failures.

This change will cause an automatic, randomized reassignment of
the bitcoind RPC and P2P ports in case it fails to come up properly.
This process retried up to four times before finally failing.

Note that this implementation tweaks the configuration file
after failure. As a further TODO item, it is likely better to
eventually rework and unify the setup and initialization phase of
the RPC tests to always present a properly initialized set of
bitcoinds instead of this approach of fixing it later on.

Most notably, this approach introduced here fails if results of
calling the function p2p_port() or rpc_port() or derivatives of these
are saved in a test case between initialization of the configuration
file and starting of the bitcoinds.

To test this commit, please tweak util.py, so that the PORT_RANGE
and debug_port_assignments settings read like this:

PORT_RANGE = 100

debug_port_assignments = True

Running the parallelized test suite with rpc-tests.py is then likely
to trigger these sporadic failures, but the retrying should recover
from these situations.

Another way to do it more specifically for one test and manually is
like this:

Run a multi-node test case, like this one:

qa/rpc-tests/abandonconflict.py --portseed=0 --nocleanup

while also, in parallel, running netcat in a second shell:

nc -l -p 11000

or

nc -l -p 16000

respectively, and observe the behavior with and without this commit.

sickpig commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -46,21 +47,56 @@
+def fixup_ports_in_configfile(i):
+    assert(i in PortSeed.config_file)
+
+    print("Tweaking ports in configuration file %s for node %d" % (PortSeed.config_file[i], i))
```
use `import logging` then define log level and logger object via:

```py
logging.basicConfig(format='%(asctime)s.%(levelname)s: %(message)s', level=logging.INFO,stream=sys.stdout)
```

then use `logging.info("...")` rather than print

sickpig reviewed (dismissed):
just skimmed through the code and it seems a good change to have. going to review further once you finish investigate the travis failure

awemany committed:
rpc-tests.py: Allow certain (assumed to be transient) errors on stderr

bitcoind will print a couple of messages on stderr when it fails to bind
to a TCP port. This makes rpc-tests.py ignore these errors when testing
for RPC test failure.

awemany committed:
Initialize logging first for all RPC tests

This initializes logging using logging.basicConfig(..) first for all RPC.
tests. This is done by importing test_framework.loginit as the first action
in any RPC test script.

Also, logging initialization was removed from those tests where it was done manually
before.

awemany committed:
Port remapping: Use logging (which is configured to stdout) instead of prints

awemany committed:
RPC tests: Add --no-ipv6-rpc-listen option and no IPV6 in travis

This option will switch off the IPv6 RPC listening socket in bitcoind
so that start up does not fail in the current travis configuration
(which does not support IPv6).

This commit also switches off the IPv6 RPC socket in travis by adding
this option to .travis.yml.

awemany committed:
rpc-tests.py: Cosmetic changes to the output

This changes the output to be a bit more readable and makes it easier
to follow which output belongs to what test.

awemany committed:
rpc-tests.py: Fix reading of stderr and stdout

This makes rpc-tests.py regularly check the state of stdout and
stderr pipes.

The test driver program did not regularly check for the status of
the stderr and stdout pipes from the respective test subprocesses.
If a tests produced a lot of output, it could have filled the output
pipe, which would lead to a hang.

Note that due to oddities in python thread handling, the .communicate()
call that waits for stdout / stderr data has to be called with a slightly
positive timeout value. A value of zero seems to stop the .join() in
python3's subprocess.py early so that the whole of rpc-tests.py can get stuck
again, simply because each communicate() attempt will fail due to the
too-low timeout!

awemany commented:
Ok, finally I found and debugging all issues with Travis interaction. Taking this out of WIP.

sickpig commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -478,9 +479,21 @@ def comms(timeout):
+
+                    # This is a list of expected messages on stderr. If they appear, they do not
+                    # necessarily indicate final failure of a test.
+                    stderr_filtered = stderr.replace("Error: Unable to start HTTP server. See debug log for details.", "")
```
I really do like code changes made in this PR. It makes the QA py suite more robust/resilient to Travis platform weirdness. The only thing that I found a little bit fragile in the PR is this part, cause we need to keep in sync the py test w/ changes that could me made to bitcoind/qt code. E.g.: 

```
src/init.cpp:924:            return InitError(_("Unable to start HTTP server. See debug log for details."));
src/qt/bitcoinstrings.cpp:396:    QT_TRANSLATE_NOOP("bitcoin-core", "Unable to start HTTP server. See debug log for details."),
```

and more to the point since those error messages are i18n enabled we risk  not detecting the error conditions if the locale/lang settings of the machine where you run the py suites run in something else the English. 

That said I approve the changes.

A quick solution is to be sure that bitcoind will be executed in an env where `LC_ALL` is set to `C`, but I guess we should add it on subsequent PR.

awemany replied:
Excellent point regarding the locale and translations!

To be honest, I didn't like the string comparison that much either but found no quick solution (and didn't think about the locale issue).

One thing one could do is to maybe prefix fatal error messages of bitcoind with some kind of error message index and create a global table of them?

I imagine bitcoind could fail with something like:

[ERR#1234] Unable to start HTTP server. See debug log for details.

and then one would just create a regexp or string match for the first [ERR #1234] part. 

I can either add that to this commit (but that will take awhile), or we take this one and I put it onto my TODO list.

What do you think? Maybe such an error table also warrants some coordination across implementations.

sickpig replied:
Think this is a fair approach even if we won't be able to achieve the use of this set of error codes on all Cash implementations.

awemany commented:
Though it looks all good on travis now, I am right now investigating an issue I have with a local RPC test being stuck and am wondering whether it is related to my code or another bug. Putting into WIP for the meanwhile.

awemany commented:
Ok, the issue seems to have been some oddity somewhere deep in python's thread handling. See the updated commit message in c3261c8.

sickpig reviewed (approved):
re approved after c3261c83a1 fixed

awemany commented:
After some further thinking, putting this back into WIP. Will do the following two further enhancements:

- Also fail on binding any of the P2P ports, as the basically same issue exists for those (the rest of the python code can stay as-is though)

- Only have bitcoind fail when a to-be-defined flag is given on the command line, as to not break any setups of current users needlessly. Maybe it would be a good idea to deprecate the 'can silently fail to bind to a port'-behavior down the road, though.

awemany committed:
Set bindallorfail option for RPC testing

awemany commented:
Improvements have been implemented, taking this out of WIP again.
